Usage doc for configuring Nomad OIDC with AWS IAM (#23845)
hashicorp/nomad: website/content/docs/operations/aws-oidc-provider.mdx

> ---

Author: @tgross does this summary make sense? I was translating a support doc initially and was a little unsure of the terminology. This is my understanding based on your explanations.

_Nomad is the OIDC provider that gets registered with AWS IAM and subsequently trusted as such. It's responsible for creating workload identity tokens (JWTs) for tasks that the job then presents to AWS IAM for federated access to AWS resources. AWS in turn validates that token with Nomad (since it's the OIDC provider)._

> ---

Author: What is the difference between this workflow and auth-method create? This one shows how to configure AWS IAM to trust Nomad as an OIDC provider whereas the other configures the actual authentication Nomad provides as the OIDC provider?

> ---

Reviewer: That summary LGTM.

Another way of looking at the difference is:
* In this workflow, AWS is configured to trust Nomad identities. That allows a Nomad workload to use AWS APIs.
* In the `auth-method create` workflow, Nomad is configured to trust AWS. That allows an AWS user to "login" to Nomad to use Nomad APIs.

> ---

Author: Ahhhh, that helps immensely

> ---

Author: Would a title like this make more sense then?
`Federate access to AWS with OIDC and Workload Identity`
Or even just
`Federate access to AWS with Nomad Workload Identity`

> ---

Reviewer: Yeah that sounds great, especially the latter.

layout: docs
page_title: Authenticate Nomad jobs with AWS IAM identity as an OIDC provider
description: |-
  Integrate Nomad as an OpenID Connect (OIDC) provider with AWS IAM Identity.
---

# Authenticate Nomad jobs with AWS IAM identity as an OIDC provider

This page describes how to integrate Nomad with AWS IAM as an OpenID Connect (OIDC) provider.
You can configure Nomad jobs to present a JSON Web Token (JWT) that the Nomad server authenticates 

> and passes to AWS IAM so that the job can access AWS services.

Author: This is the kind of clarification I needed, thanks a bunch - I'll update it to be more clear

resource "aws_lb" "test" {
  name               = "test-lb-tf"
  internal           = false
  load_balancer_type = "application"
  security_groups    = [<SECURITY_GROUP_ID>]
  subnets            = [<SUBNET_IDS>]

  enable_deletion_protection = true

  access_logs {
    bucket  = <S3_BUCKET_ID>
    prefix  = "test-lb"
    enabled = true
  }
}
```

This configuration requires the following information:
- [`<SECURITY_GROUP_ID>`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb#security_groups):
A list of security groups IDs that you want to apply to the load balancer.
- [`<SUBNET_IDS>`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb#subnets):
A list of subnet IDs that you want to attach to the load balancer.
- [`<S3_BUCKET_ID>`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb#bucket):
The ID of an S3 bucket to store the load balancer access logs in.

### Create a load balancer listener

Use the [`aws_lb_listener` resource](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lb_listener).

```hcl
resource "aws_lb_listener" "example" {
  load_balancer_arn = <LB_ARN>

>   port              = "443"

Reviewer: The Nomad agents are listening on port 4646, but I don't see how we're mapping port 443 to that?

Reviewer: I'm not sure which "original guide" you're referring to here, but I think I also assumed this had been verified end-to-end and not just ported from some other source.

Reviewer: Ok, maybe we should pull in @SuyashHashiCorp to take a look at this doc as well then? If I recall correctly that guide may have also been inspired by work @schmichael did.

Reviewer 2: I'm really excited to see workload identity federation (WIF) documentation for another cloud provider, but I'm afraid this content probably needs to be moved to https://github.com/hashicorp/tutorials if it is intended to be complementary to our GCP WIF tutorial: https://developer.hashicorp.com/nomad/tutorials/fed-workload-identity/integration-gcp

Source for that is here https://github.com/hashicorp/tutorials/blob/main/content/tutorials/nomad/integration-gcp.mdx and the corresponding jobspecs/terraform are here: https://github.com/hashicorp-education/learn-nomad-workload-identity-federation/tree/main

[proxy.nomad.hcl](https://github.com/hashicorp-education/learn-nomad-workload-identity-federation/blob/main/proxy.nomad.hcl) is what handles proxying JWKS requests from the cloud vendor to Nomad's HTTP endpoint. You should be able to use that jobspec for any deployment in the cloud or onprem. The flow in GCP is:

```
GCP  ---https (443)---> LB ---http (80)---> Proxy ---task api (unix socket)---> Nomad API
```

...and I assume a similar flow works for any cloud.

---

If there is a desire to keep this content in docs instead of tutorials, I think we should move the GCP tutorial over here and create a subdirectory under `Operations` for both. Ideally we'd add more WIF examples in the future. This all _feels_ more like tutorials to me, but I can understand the desire and simplicity of reference docs. If we have any hard and fast rules about what makes something a tutorial vs a doc, I don't know it, so as long as we're consistent I'm ok with WIF content in either place as long as we're consistent.

Author: @schmichael that's totally valid. Education is working on a larger initiative to better organize tutorials and docs and we've got some ongoing plans. @boruszak or I can loop you in for better context but the gist is: there are lots of tutorials/docs that could and should be moved around and edited to better fit in their respective places. 

This doc was being brought over from the support team's knowledge base to make it more visible and we did intend to shuffle it around as part of the larger edu goal when we get there. 😅

[`oidc_issuer` attribute](/nomad/docs/configuration/server#oidc_issuer) and set the value
to the domain name for the Nomad cluster. This [enables the HTTP endpoint in Nomad](https://developer.hashicorp.com/nomad/docs/configuration/server#oidc_issuer) that allows third parties to discover Nomad's OIDC
configuration.

```hcl
server {
  enabled = true
  [...]
  oidc_issuer = "https://<DOMAIN_NAME>"
  [...]
}
```

This configuration requires the following information:
- `<DOMAIN_NAME>`: The domain name of your Nomad cluster without the protocol or port.

Restart the Nomad server agent to apply the configuration changes.

## Create and run a sample jobspec file

Create and run a jobspec file to validate your configuration. The following file is named `s3-upload.nomad.hcl`, add the following configuration to it, and
save the file.

```hcl
job "s3" {
  type = "batch"
  group "bucket" {
    task "copy" {
      driver = "docker"
      config {
        image = "public.ecr.aws/aws-cli/aws-cli"
        command = "s3"
        args = ["cp", "/local/test.txt", "s3://<S3_BUCKET_NAME>/test-nomad.txt"]
      }

      identity {
        name = "aws"
        aud = ["aws"] 
        file = true

>         env = true

Reviewer: We probably don't need `env = true`, as we're setting a specific environment variable `AWS_WEB_IDENTITY_TOKEN_FILE` from the `file = true` instead.

        ttl = "1h" 
        change_mode = "restart"
      }

      template {
        destination = "local/test.txt"
        change_mode = "restart"
        data        = <<EOF
Job:          {{ env "NOMAD_JOB_NAME" }}
Alloc:        {{ env "NOMAD_ALLOC_ID" }}
EOF
      }

      env {
        AWS_ROLE_ARN = "arn:aws:iam::<AWS_ACCOUNT_ID>:role/<IAM_ROLE_NAME>"
        # The format of the token file is nomad_$NAME_OF_IDENTITY.jwt
        AWS_WEB_IDENTITY_TOKEN_FILE = "${NOMAD_SECRETS_DIR}/nomad_aws.jwt"
      }

      resources {
        cpu    = 500
        memory = 256
      }
    }
  }
}
```

This configuration requires the following information:
- `<S3_BUCKET_NAME>`: The name of the S3 bucket where the test file will be saved.
- `<AWS_ACCOUNT_ID>`: The ID of the AWS account where the IAM role from the previous step was
created.
- `<IAM_ROLE_NAME>`: The name of the IAM role from previous steps. This will be

> `s3_all_access_role` if you are using the examples from this page.

Reviewer: We should probably have these as [HCL2 Variables](https://developer.hashicorp.com/nomad/docs/job-specification/hcl2/variables)?

Author: Ideally yes, it's intended to be a usage guide and not a tutorial though so I wanted to keep the snippets minimal so it's easier to quickly parse

The zone ID of the load balancer. This will be `aws_lb.test.zone_id` if you are using the
examples from this page.

### Create an OIDC Identity Provider

Use the [`aws_iam_openid_connect_provider` resource](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_openid_connect_provider).

```hcl
data "tls_certificate" "example" {
  url = <CERT_DOMAIN_NAME>
}

resource "aws_iam_openid_connect_provider" "nomad" {
  # Nomad HTTPS URL
  url = <CERT_DOMAIN_NAME>

  client_id_list = [
    "aws",
  ]

  thumbprint_list = [data.tls_certificate.example.certificates.0.sha1_fingerprint]
}
```

This configuration requires the following information:
- [`<CERT_DOMAIN_NAME>`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_openid_connect_provider#url):
The domain name of the load balancer certificate. This will be
`aws_acm_certificate.example.domain_name` if you are using the examples from this page.

### Create an IAM policy for OIDC Federated Users

Use the [`aws_iam_role` resource](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role)

> to create an appropriate IAM role for federated users. This will be specific to your use case.
> The following example sets S3 bucket access.

Reviewer: Although the AWS docs use "users", we're always talking about workloads here, so clarifying that might be helpful:

```suggestion
to create an appropriate IAM role for workloads acting as federated users. This will be 
specific to your use case. The following example allows workloads access to S3 buckets.
```

---
layout: docs

> page_title: Authenticate Nomad jobs with AWS IAM Identity Provider

Author: ```suggestion
page_title: Authenticate Nomad jobs with AWS IAM identity as an OIDC provider
```
Did a bunch more reading and found that AWS IAM Identity Provider isn't actually a product. Also Nomad is the provider here so wanted to clarify that

description: |-
  Integrate Nomad as an OpenID Connect (OIDC) provider with AWS IAM Identity.
---

# Authenticate Nomad jobs with AWS IAM Identity Provider

> This page describes how to integrate Nomad with AWS IAM Identity Provider as an OpenID Connect
> (OIDC). You can configure Nomad jobs to present a JWT token that the Nomad server authenticates 
> with AWS IAM before it executes a task.

Author: ```suggestion
This page describes how to integrate Nomad with AWS IAM as an OpenID Connect (OIDC) provider.
You can configure Nomad jobs to present a JSON Web Token (JWT) that the Nomad server authenticates 
and passes to AWS IAM so that the job can access AWS services.
```

## Prerequisites

> To integrate Nomad and the AWS IAM Identity Provider, you need a running Nomad cluster that
> meets the following prerequisites:

Author: ```suggestion
To integrate Nomad and AWS IAM identity, you need a running Nomad cluster that
meets the following prerequisites:
```

- Nomad v1.7.x or later
- TLS enabled

The instructions on this page also assume the following:

- Your AWS account has the necessary permissions to create IAM roles, policies, hosted zones,
  and certificates.
- Terraform is installed locally and 
  [configured to communicate with AWS](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#authentication-and-configuration).

## Workflow

> The process to integrate Nomad with the AWS IAM provider consists of the following steps:

Author: ```suggestion
The process to integrate Nomad with AWS IAM identity consists of the following steps:
```

> 1. Create the required AWS resources:
>   - [Route 53 hosted zone](#create-a-hosted-zone)
>   - [AWS Certificate Manager SSL certificates](#generate-ssl-certificates)
>   - [Load balancer and listener](#create-an-application-load-balancer)
>   - [DNS Alias for the load balancer](#create-a-dns-alias)
>   - [OIDC Identity Provider](#create-an-oidc-identity-provider)
>   - [AWS IAM role](#create-an-iam-policy-for-oidc-federated-users)

Author: ```suggestion
1. Create the required AWS resources:
    - [Route 53 hosted zone](#create-a-hosted-zone)
    - [AWS Certificate Manager SSL certificates](#generate-ssl-certificates)
    - [Load balancer and listener](#create-an-application-load-balancer)
    - [DNS Alias for the load balancer](#create-a-dns-alias)
    - [OIDC Identity Provider](#create-an-oidc-identity-provider)
    - [AWS IAM role](#create-an-iam-policy-for-oidc-federated-users)
```
Extra spacing so the numbered list works correctly

1. Update the Nomad server configuration.

> 1. Create a jobspec file that uses AWS OIDC and then run it to verify your configuration.

Author: ```suggestion
1. Create a jobspec file that accesses AWS and then run it to verify your configuration.
```
